https://invent.kde.org/frameworks/kwayland/-/merge_requests/61

From e9449bad15f3f62771105a517e53e0b2830514b8 Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fabian@ritter-vogt.de>
Date: Fri, 29 Apr 2022 12:23:31 +0200
Subject: [PATCH] PlasmaWindowManagement: Avoid unbounded recursion and delay
 in readData

It recurses for each partial read, which can lead to unbounded recursion depth
and delay (retryCount isn't passed). Convert the code to be iterative to fix
that, which is also much simpler.

BUG: 453124
---
 src/client/plasmawindowmanagement.cpp | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/src/client/plasmawindowmanagement.cpp b/src/client/plasmawindowmanagement.cpp
index 2a3c608..fe893f8 100644
--- a/src/client/plasmawindowmanagement.cpp
+++ b/src/client/plasmawindowmanagement.cpp
@@ -663,16 +663,14 @@ static int readData(int fd, QByteArray &data)
     int n;
     while (true) {
         n = QT_READ(fd, buf, sizeof buf);
-        if (n == -1 && (errno == EAGAIN) && ++retryCount < 1000) {
+        if (n > 0) {
+            data.append(buf, n);
+        } else if (n == -1 && (errno == EAGAIN) && ++retryCount < 1000) {
             usleep(1000);
         } else {
             break;
         }
     }
-    if (n > 0) {
-        data.append(buf, n);
-        n = readData(fd, data);
-    }
     return n;
 }
 
-- 
GitLab

